File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/gameroom/GameRoomStoreRecordSupport.kt
package com.pillyliu.pinprofandroid.gameroom

internal data class GameRoomIssueMutationResult(
    val state: GameRoomPersistedState,
    val issueId: String,
)

internal fun gameRoomStateWithAddedEvent(
    state: GameRoomPersistedState,
    machineID: String,
    type: MachineEventType,
    category: MachineEventCategory,
    summary: String,
    occurredAtMs: Long?,
    notes: String?,
    partsUsed: String?,
    consumablesUsed: String?,
    pitchValue: Double?,
    pitchMeasurementPoint: String?,
    playCountAtEvent: Int?,
    linkedIssueID: String?,
    now: Long,
): Pair<GameRoomPersistedState, String> {
    val occurredAt = occurredAtMs ?: now
    val event = MachineEvent(
        ownedMachineID = machineID,
        type = type,
        category = category,
        occurredAtMs = occurredAt,
        playCountAtEvent = playCountAtEvent,
        summary = summary.trim().ifBlank { "Event" },
        notes = normalizeOptionalGameRoomString(notes),
        partsUsed = normalizeOptionalGameRoomString(partsUsed),
        consumablesUsed = normalizeOptionalGameRoomString(consumablesUsed),
        pitchValue = pitchValue,
        pitchMeasurementPoint = normalizeOptionalGameRoomString(pitchMeasurementPoint),
        linkedIssueID = normalizeOptionalGameRoomString(linkedIssueID),
        createdAtMs = now,
        updatedAtMs = now,
    )
    return state.copy(events = state.events + event) to event.id
}

internal fun gameRoomStateWithUpdatedEvent(
    state: GameRoomPersistedState,
    id: String,
    occurredAtMs: Long,
    summary: String,
    notes: String?,
    now: Long,
): GameRoomPersistedState {
    return state.copy(
        events = state.events.map { event ->
            if (event.id != id) return@map event
            event.copy(
                occurredAtMs = occurredAtMs,
                summary = summary.trim().ifBlank { "Event" },
                notes = normalizeOptionalGameRoomString(notes),
                updatedAtMs = now,
            )
        },
    )
}

internal fun gameRoomStateWithDeletedEvent(
    state: GameRoomPersistedState,
    id: String,
): GameRoomPersistedState {
    return state.copy(
        events = state.events.filterNot { it.id == id },
        attachments = state.attachments.filterNot {
            it.ownerType == MachineAttachmentOwnerType.event && it.ownerID == id
        },
    )
}

internal fun gameRoomAttachmentsForMachine(
    state: GameRoomPersistedState,
    machineID: String,
): List<MachineAttachment> {
    return state.attachments
        .filter { it.ownedMachineID == machineID }
        .sortedByDescending { it.createdAtMs }
}

internal fun gameRoomAttachmentsForEvent(
    state: GameRoomPersistedState,
    eventID: String,
): List<MachineAttachment> {
    return state.attachments
        .filter { it.ownerType == MachineAttachmentOwnerType.event && it.ownerID == eventID }
        .sortedByDescending { it.createdAtMs }
}

internal fun gameRoomAttachmentsForIssue(
    state: GameRoomPersistedState,
    issueID: String,
): List<MachineAttachment> {
    return state.attachments
        .filter { it.ownerType == MachineAttachmentOwnerType.issue && it.ownerID == issueID }
        .sortedByDescending { it.createdAtMs }
}

internal fun gameRoomStateWithAddedAttachment(
    state: GameRoomPersistedState,
    machineID: String,
    ownerType: MachineAttachmentOwnerType,
    ownerID: String,
    kind: MachineAttachmentKind,
    uri: String,
    thumbnailURI: String?,
    caption: String?,
): Pair<GameRoomPersistedState, String>? {
    val normalizedURI = uri.trim()
    if (normalizedURI.isBlank()) return null
    val attachment = MachineAttachment(
        ownedMachineID = machineID,
        ownerType = ownerType,
        ownerID = ownerID,
        kind = kind,
        uri = normalizedURI,
        thumbnailURI = normalizeOptionalGameRoomString(thumbnailURI),
        caption = normalizeOptionalGameRoomString(caption),
    )
    return state.copy(attachments = state.attachments + attachment) to attachment.id
}

internal fun gameRoomStateWithDeletedAttachment(
    state: GameRoomPersistedState,
    id: String,
): GameRoomPersistedState {
    return state.copy(attachments = state.attachments.filterNot { it.id == id })
}

internal fun gameRoomStateWithUpdatedAttachment(
    state: GameRoomPersistedState,
    id: String,
    caption: String?,
    notes: String?,
    now: Long,
): GameRoomPersistedState? {
    val attachment = state.attachments.firstOrNull { it.id == id } ?: return null
    return state.copy(
        attachments = state.attachments.map { current ->
            if (current.id != id) current else current.copy(caption = normalizeOptionalGameRoomString(caption))
        },
        events = if (attachment.ownerType == MachineAttachmentOwnerType.event) {
            state.events.map { event ->
                if (event.id != attachment.ownerID) event else event.copy(
                    notes = normalizeOptionalGameRoomString(notes),
                    updatedAtMs = now,
                )
            }
        } else {
            state.events
        },
    )
}

internal fun gameRoomStateWithDeletedAttachmentAndLinkedEvent(
    state: GameRoomPersistedState,
    id: String,
): GameRoomPersistedState? {
    val attachment = state.attachments.firstOrNull { it.id == id } ?: return null
    var nextAttachments = state.attachments.filterNot { it.id == id }
    var nextEvents = state.events
    if (attachment.ownerType == MachineAttachmentOwnerType.event) {
        nextEvents = nextEvents.filterNot { it.id == attachment.ownerID }
        nextAttachments = nextAttachments.filterNot {
            it.ownerType == MachineAttachmentOwnerType.event && it.ownerID == attachment.ownerID
        }
    }
    return state.copy(
        attachments = nextAttachments,
        events = nextEvents,
    )
}

internal fun gameRoomStateWithOpenedIssue(
    state: GameRoomPersistedState,
    machineID: String,
    symptom: String,
    severity: MachineIssueSeverity,
    subsystem: MachineIssueSubsystem,
    openedAtMs: Long?,
    diagnosis: String?,
    now: Long,
): GameRoomIssueMutationResult {
    val openedAt = openedAtMs ?: now
    val issue = MachineIssue(
        ownedMachineID = machineID,
        status = MachineIssueStatus.open,
        severity = severity,
        subsystem = subsystem,
        symptom = symptom.trim().ifBlank { "Issue" },
        diagnosis = normalizeOptionalGameRoomString(diagnosis),
        openedAtMs = openedAt,
        createdAtMs = now,
        updatedAtMs = now,
    )
    val (nextState, _) = gameRoomStateWithAddedEvent(
        state = state.copy(issues = state.issues + issue),
        machineID = machineID,
        type = MachineEventType.issueOpened,
        category = MachineEventCategory.issue,
        summary = "Issue opened: ${issue.symptom}",
        occurredAtMs = openedAt,
        notes = diagnosis,
        partsUsed = null,
        consumablesUsed = null,
        pitchValue = null,
        pitchMeasurementPoint = null,
        playCountAtEvent = null,
        linkedIssueID = issue.id,
        now = now,
    )
    return GameRoomIssueMutationResult(
        state = nextState,
        issueId = issue.id,
    )
}

internal fun gameRoomStateWithResolvedIssue(
    state: GameRoomPersistedState,
    issueID: String,
    resolution: String?,
    resolvedAtMs: Long?,
    now: Long,
): GameRoomPersistedState? {
    val issue = state.issues.firstOrNull { it.id == issueID } ?: return null
    val resolvedAt = resolvedAtMs ?: now
    val issueState = state.copy(
        issues = state.issues.map { current ->
            if (current.id != issueID) return@map current
            current.copy(
                status = MachineIssueStatus.resolved,
                resolvedAtMs = resolvedAt,
                resolution = normalizeOptionalGameRoomString(resolution),
                updatedAtMs = now,
            )
        },
    )
    val (nextState, _) = gameRoomStateWithAddedEvent(
        state = issueState,
        machineID = issue.ownedMachineID,
        type = MachineEventType.issueResolved,
        category = MachineEventCategory.issue,
        summary = "Issue resolved: ${issue.symptom}",
        occurredAtMs = resolvedAt,
        notes = resolution,
        partsUsed = null,
        consumablesUsed = null,
        pitchValue = null,
        pitchMeasurementPoint = null,
        playCountAtEvent = null,
        linkedIssueID = issueID,
        now = now,
    )
    return nextState
}

File: Pinball App 2/Pinball App 2/gameroom/GameRoomStoreRecordSupport.swift
import Foundation

extension GameRoomStore {
    func updateEvent(id: UUID, occurredAt: Date, summary: String, notes: String?) {
        guard let index = state.events.firstIndex(where: { $0.id == id }) else { return }
        state.events[index].occurredAt = occurredAt
        state.events[index].summary = normalizedOptionalString(summary) ?? "Event"
        state.events[index].notes = normalizedOptionalString(notes)
        state.events[index].updatedAt = Date()
        saveAndRecompute()
    }

    func deleteEvent(id: UUID) {
        state.events.removeAll { $0.id == id }
        state.attachments.removeAll { $0.ownerType == .event && $0.ownerID == id }
        saveAndRecompute()
    }

    @discardableResult
    func addEvent(
        machineID: UUID,
        type: MachineEventType,
        category: MachineEventCategory,
        occurredAt: Date = Date(),
        playCountAtEvent: Int? = nil,
        summary: String,
        notes: String? = nil,
        partsUsed: String? = nil,
        consumablesUsed: String? = nil,
        pitchValue: Double? = nil,
        pitchMeasurementPoint: String? = nil,
        linkedIssueID: UUID? = nil
    ) -> UUID {
        let now = Date()
        let event = MachineEvent(
            ownedMachineID: machineID,
            type: type,
            category: category,
            occurredAt: occurredAt,
            playCountAtEvent: playCountAtEvent,
            summary: normalizedOptionalString(summary) ?? "Event",
            notes: normalizedOptionalString(notes),
            partsUsed: normalizedOptionalString(partsUsed),
            consumablesUsed: normalizedOptionalString(consumablesUsed),
            pitchValue: pitchValue,
            pitchMeasurementPoint: normalizedOptionalString(pitchMeasurementPoint),
            linkedIssueID: linkedIssueID,
            createdAt: now,
            updatedAt: now
        )
        state.events.append(event)
        saveAndRecompute()
        return event.id
    }

    @discardableResult
    func openIssue(
        machineID: UUID,
        openedAt: Date = Date(),
        symptom: String,
        severity: MachineIssueSeverity = .medium,
        subsystem: MachineIssueSubsystem = .other,
        diagnosis: String? = nil
    ) -> UUID {
        let now = Date()
        let issue = MachineIssue(
            ownedMachineID: machineID,
            status: .open,
            severity: severity,
            subsystem: subsystem,
            symptom: normalizedOptionalString(symptom) ?? "Issue",
            diagnosis: normalizedOptionalString(diagnosis),
            openedAt: openedAt,
            createdAt: now,
            updatedAt: now
        )
        state.issues.append(issue)
        saveAndRecompute()
        return issue.id
    }

    func resolveIssue(id: UUID, resolvedAt: Date = Date(), resolution: String?) {
        guard let index = state.issues.firstIndex(where: { $0.id == id }) else { return }
        state.issues[index].status = .resolved
        state.issues[index].resolvedAt = resolvedAt
        state.issues[index].resolution = normalizedOptionalString(resolution)
        state.issues[index].updatedAt = Date()
        saveAndRecompute()
    }

    func addAttachment(
        machineID: UUID,
        ownerType: MachineAttachmentOwnerType,
        ownerID: UUID,
        kind: MachineAttachmentKind,
        uri: String,
        caption: String? = nil
    ) {
        let normalizedURI = normalizedOptionalString(uri)
        guard let normalizedURI else { return }
        state.attachments.append(
            MachineAttachment(
                ownedMachineID: machineID,
                ownerType: ownerType,
                ownerID: ownerID,
                kind: kind,
                uri: normalizedURI,
                caption: normalizedOptionalString(caption)
            )
        )
        saveAndRecompute()
    }

    func updateAttachment(
        id: UUID,
        caption: String?,
        notes: String?
    ) {
        guard let attachmentIndex = state.attachments.firstIndex(where: { $0.id == id }) else { return }
        state.attachments[attachmentIndex].caption = normalizedOptionalString(caption)

        let attachment = state.attachments[attachmentIndex]
        if attachment.ownerType == .event,
           let eventIndex = state.events.firstIndex(where: { $0.id == attachment.ownerID }) {
            state.events[eventIndex].notes = normalizedOptionalString(notes)
            state.events[eventIndex].updatedAt = Date()
        }

        saveAndRecompute()
    }

    func deleteAttachmentAndLinkedEvent(id: UUID) {
        guard let attachment = state.attachments.first(where: { $0.id == id }) else { return }
        state.attachments.removeAll { $0.id == id }

        if attachment.ownerType == .event {
            state.events.removeAll { $0.id == attachment.ownerID }
            state.attachments.removeAll { $0.ownerType == .event && $0.ownerID == attachment.ownerID }
        }

        saveAndRecompute()
    }
}
